src/time_source.rs

```diff
@@ -0,0 +1,89 @@
+//! Module containing Aerugo's time source module, providing configurable timestamps for the system
+//! Should be used internally by the system.
+
+use aerugo_hal::system_hal::SystemHal;
+
+use crate::hal::Hal;
+use crate::internal_cell::InternalCell;
+use crate::{Duration, Instant};
+
+/// Time source, responsible for creating timestamps.
+///
+/// Allows time tracking/timestamp generation since three points in time:
+/// * Hardware initialization (call to [`Aerugo::initialize`](crate::Aerugo::initialize))
+/// * Start of Aerugo scheduler (call to [`Aerugo::start`](crate::Aerugo::start))
+/// * User-defined offset
+pub struct TimeSource {
+    /// Time since system's scheduler start.
+    system_start_offset: InternalCell<Option<Duration>>,
+    /// User-defined offset.
+    user_offset: InternalCell<Option<Duration>>,
+}
+
+impl TimeSource {
+    /// Creates new instance of TimeSource
+    pub const fn new() -> Self {
+        TimeSource {
+            system_start_offset: InternalCell::new(None),
+            user_offset: InternalCell::new(None),
+        }
+    }
+
+    /// Returns time since system initialization (call to [`Aerugo::initialize`](crate::Aerugo::initialize), start of the hardware timer)
+    pub fn time_since_init() -> Instant {
+        Hal::get_system_time()
+    }
+
+    /// Returns time since system's scheduler start (call to [`Aerugo::start`](crate::Aerugo::start)), or `None` if system hasn't started yet.
+    pub fn time_since_start(&self) -> Option<Instant> {
+        // SAFETY: This is safe as long as used in single-core context and `mark_system_start` is not being called.
+        if let Some(start_offset) = unsafe { *self.system_start_offset.as_ref() } {
```

Comment:
imo `match` would be more readable here

Reply:
fix'd, also made better safety docs for TimeSource